packages/micropip/test_micropip.py

```diff
@@ -74,32 +74,44 @@ def __init__(self):
         self.releases_map = {}
         self.metadata_map = {}
 
-    def add_pkg(
+    def add_pkg_version(
         self,
         name: str,
-        requirements: dict[str, list[str]],
+        version: str = "1.0.0",
+        *,
+        requirements: list[str] | None = None,
+        extras: dict[str, list[str]] | None = None,
         platform: str = "generic",
     ):
-        releases = {}
-        for version, reqs in requirements.items():
-            filename = make_wheel_filename(name, version, platform)
-            releases[version] = [
-                {
-                    "filename": filename,
-                    "url": filename,
-                    "digests": {
-                        "sha256": Wildcard(),
-                    },
-                }
+        if requirements is None:
+            requirements = []
+        if extras is None:
+            extras = {}
+        if name not in self.releases_map:
+            self.releases_map[name] = {"releases": {}}
+        releases = self.releases_map[name]["releases"]
+        filename = make_wheel_filename(name, version, platform)
+        releases[version] = [
+            {
+                "filename": filename,
+                "url": filename,
+                "digests": {
+                    "sha256": Wildcard(),
+                },
+            }
+        ]
+        metadata = [("Name", name), ("Version", version)] + [
+            ("Requires-Dist", req) for req in requirements
+        ]
+        for extra, reqs in extras.items():
+            metadata += [("Provides-Extra", extra)] + [
```

Comment:
@hoodmane I am just catching up on this, and to learn more. Where can I find a description of this metadata specification which explains what `Provides-Extra` is? I understand what it is practically, but would like to read a definitive source on it.

Reply:
> @hoodmane I am just catching up on this, and to learn more. Where can I find a description of this metadata specification which explains what `Provides-Extra` is? I understand what it is practically, but would like to read a definitive source on it.

Never mind i think i found some reference in the pip source code, i guess i have a good starting point there. thanks.

Reply:
The other thing I recommend is looking at the actual `METADATA` files for installed wheels in your system.